sympy/polys/domains/fpsring.py

```diff
@@ -0,0 +1,522 @@
+
+        Examples
+        ========
+        >>> from sympy import QQ
+        >>> from sympy.polys.domains.fpsring import PowerSeriesRing
+        >>> R = PowerSeriesRing(QQ, 'x', 10)
+        >>> x = R.gen
+        >>> f = 1 + 2*x + 3*x**2
+        >>> g =  5 + 6*x + x**4
+        >>> f * g
+        5 + 16*x + 27*x**2 + 18*x**3 + x**4 + 2*x**5 + 3*x**6 + O(x**10)
+        """
+        if isinstance(other, PowerSeriesElement):
+            if self.ring != other.ring:
+                raise ValueError("Cannot multiply PowerSeriesElement from different domains")
+            return self._mul(other)
+
+        domain: Domain = self.ring.domain
+        if isinstance(other, int):
+            return self._mul_ground(domain.convert_from(QQ(other), QQ))
+        elif domain.of_type(other):
+            return self._mul_ground(other)
+        else:
+            raise NotImplementedError
+
+    def __rmul__(self, other: int | Any) -> PowerSeriesElement:
+        """Multiply a scalar by this power series (right multiplication).
+
+        Examples
+        ========
+        >>> from sympy import QQ
+        >>> from sympy.polys.domains.fpsring import PowerSeriesRing
+        >>> R = PowerSeriesRing(QQ, 'x', 10)
+        >>> x = R.gen
+        >>> f =  1 + 2*x + 3*x**2
+        >>> 5 * f
+        5 + 10*x + 15*x**2 + O(x**10)
+        """
+        domain: Domain = self.ring.domain
+        if isinstance(other, int):
+            return self._mul_ground(domain.convert_from(QQ(other), QQ))
+        elif domain.of_type(other):
+            return self._mul_ground(other)
+        else:
+            raise NotImplementedError
+
+    def __pow__(self, n: int) -> PowerSeriesElement:
+        """Raise this power series to an `n` power.
+
+        Examples
+        ========
+        >>> from sympy import QQ
+        >>> from sympy.polys.domains.fpsring import PowerSeriesRing
+        >>> R = PowerSeriesRing(QQ, 'x', 10)
+        >>> x = R.gen
+        >>> f =  1 + x + x**4
+        >>> f**2
+        1 + 2*x + x**2 + 2*x**4 + 2*x**5 + x**8 + O(x**10)
+        """
+        if not isinstance(n, int):
+            raise ValueError("Exponent must be an integer.")
+        if n < 0:
+            raise ValueError("Negative exponent is not supported for power series.")
+        return self._pow(n)
+
+    def _add(self, other: PowerSeriesElement) -> PowerSeriesElement:
+        new_series = dup_add(self.poly.to_list(),
+                             other.poly.to_list(), self.ring.domain)
+        return self._new(new_series)
+
+    def _add_ground(self, other: PowerSeriesElement) -> PowerSeriesElement:
+        new_series = dup_add_ground(self.poly.to_list(),
+                                    other, self.ring.domain)
+        return self._new(new_series)
+
+    def _sub(self, other: PowerSeriesElement) -> PowerSeriesElement:
+        new_series = dup_sub(self.poly.to_list(),
+                             other.poly.to_list(), self.ring.domain)
```

Comment:
Every function is converting with `to_list`. If we want to use the `dup_*` functions then it is better to work with the `dup` representation directly i.e. just use a list of domain elements rather than DMP or DMP_Python.